medmodels/treatment_effect_estimation/builder.py

```diff
@@ -0,0 +1,312 @@
+                If None, the duration is left as it was. Defaults to 0.
+            reference (Optional[Literal['first', 'last']], optional): The reference
+                point for the grace period. Must be either 'first' or 'last'.
+                Defaults to None.
+
+        Returns:
+            TreatmentEffectBuilder: The current instance of the TreatmentEffectBuilder
+                with updated time attribute.
+        """
+        if days is not None:
+            self.grace_period_days = days
+
+        if reference is not None:
+            self.grace_period_reference = reference
+
+        return self
+
+    def set_follow_up_period(
+        self,
+        days: Optional[int] = None,
+        reference: Optional[Literal["first", "last"]] = None,
+    ) -> TreatmentEffectBuilder:
+        """Sets the follow-up period for the treatment effect estimation.
+
+        Args:
+            days (Optional[int], optional): The duration of the follow-up period
+                in days. If None, the duration is left as it was. Defaults to 365.
+            reference (Optional[Literal['first', 'last']], optional): The reference
+                point for the follow-up period. Must be either 'first' or 'last'.
+                Defaults to None.
+
+        Returns:
+            TreatmentEffectBuilder: The current instance of the TreatmentEffectBuilder
+                with updated time attribute.
+        """
+        if days is not None:
+            self.follow_up_period_days = days
+
+        if reference is not None:
+            self.follow_up_period_reference = reference
+
+        return self
+
+    def set_outcome_before_treatment_exclusion(
+        self, days: int
+    ) -> TreatmentEffectBuilder:
+        """Define whether we allow the outcome to exist before the treatment or not.
+        The outcome_before_treatment_days parameter is used to set the number of days
+        before the treatment that the outcome should not exist. If not set, the outcome
+        is allowed to exist before the treatment.
+
+        Args:
+            days (int): The number of days before the treatment that the outcome should
+                not exist.
+
+        Returns:
+            TreatmentEffectBuilder: The current instance of the TreatmentEffectBuilder
+                with updated time attribute.
+        """
+        self.outcome_before_treatment_days = days
+
+        return self
+
+    def filter_controls(self, operation: NodeOperation) -> TreatmentEffectBuilder:
+        """Filter the control group based on the provided operation.
+
+        Args:
+            operation (NodeOperation): The operation to be applied to the control group.
+
+        Returns:
+            TreatmentEffectBuilder: The current instance of the TreatmentEffectBuilder
+                with updated time attribute.
+        """
+        self.filter_controls_operation = operation
+
+        return self
+
+    def adjust_with_propensity_matching(
```

Comment:
would probably rename this to `with_propensity_adjustment` but have no hard feelings about it. If you feel like this is more concise then thats fine by me